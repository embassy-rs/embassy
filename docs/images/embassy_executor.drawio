<mxfile host="app.diagrams.net" modified="2021-12-10T08:37:46.039Z" agent="5.0 (Macintosh)" etag="F30pje_XkDZMiNl31GPU" version="15.9.4" type="device"><diagram id="6eJnfHJiGOpK0Luz4X_S" name="Page-1">7Vnhb6IwFP9r+LgFKFX8OJ13u2S3LLcld98uFQo0Q2qgTt1ff2UUsa0IU5y6XLIs9PX1WX7v1997VQOMpsvvKZpFP6mPY8M2/aUBbg3bHjg9/j83rAoDdPuFIUyJX5isyvBE3rAwmsI6Jz7OJEdGaczITDZ6NEmwxyQbSlO6kN0CGsufOkMh1gxPHop162/is6iwutCs7HeYhFH5yZYpZqaodBaGLEI+XWyYwNgAo5RSVjxNlyMc59iVuBTrvtXMrjeW4oS1WfDAsjuIh89/fv8NyGhxv0zMH1dOEeUVxXPxwuMl9uaMpmLTbFUikdJ54uM8mGWA4SIiDD/NkJfPLnjquS1i01hMZy+YeZEYBCSORzTmMfNAIHA97Hm5E0vpC96YmbjQ4ciCodgVThle1r6utQaRkw/TKWbpiruUC3oC95UyXlRptMrcRBspLP2QYE64Dl2Byx8Evh/AGmpYP6PshVus42HtI+wGW7HueS6eBN1g7djnhnVPw1rDGCf+TS4QfOTFKMuIJ8Na5cDsCmTsa2KjQMx3SOeph5uPLENpiFkT3fSUpThGjLzK+9iGv1j6SAnf4TrVoC+nGlhKCov9i1WbcqQEcpoCFS+oBXqnw/p99mdIX2OIdc3HjzRfqlKF5+0eTXhl47TAGXlDk/cpU2YMikmY5HTiOcWcAcP8UBFeS27ExJT4fr5QZpNUwETkqmw0HMx69h+Q+oOOnlsnc/oRvLCSosqc3T+1zNl6/RZgP3QI9hGgA4OTQ/exEpHQJD+4PsqiNYbtikVb6dex28AGboGmtB2o6I6rpKa/r6KrgZzPVXQAWmSUp+FJDDMcTnkKxpVp+LGeQCJDJ/LUSYPgtmwQnJOyzlTIYu7LOqiIMmzHOp5ntNpwm+UOWf2Ge9b2DdftS/Uf7Ha37V3u/KHYb7cHRq8ezroHAjf87xdG/mrrISrbob3aH7WHEtjz4HBowNtdJaeuTWrsiQp1qC1QV+a12Td7Ug6sw5heutAgyPBxFE+/Up6J4n3ulQheguJB2HDhaat4XHpkmqoFuyPFU+u5ZTVIGNzpfyQN09s4W9awR05xkoRfRMVqbvUXrWL6jfFUKtahJrX9msY+qShZqijt2/zbaj9nH0eU1D7JsT9DZAYaQ2GrL4suR1fcJl2xXXjuQlLeN89ASE56AWwrPTU5/38BbNXeKBfA4yiPY2mUBl9LeRzT2KU85jXouRLuV/bpdIgPqx9sC/fqV28w/gc=</diagram></mxfile>