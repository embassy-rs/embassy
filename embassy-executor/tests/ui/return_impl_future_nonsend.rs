#![cfg_attr(feature = "nightly", feature(impl_trait_in_assoc_type))]

use core::future::Future;

use embassy_executor::SendSpawner;

#[embassy_executor::task]
fn task() -> impl Future<Output = ()> {
    // runs in spawning thread
    let non_send: *mut () = core::ptr::null_mut();
    async move {
        // runs in executor thread
        println!("{}", non_send as usize);
    }
}

fn send_spawn(s: SendSpawner) {
    s.spawn(task()).unwrap();
}

fn main() {}
