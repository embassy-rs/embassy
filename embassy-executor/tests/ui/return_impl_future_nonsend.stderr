error: future cannot be sent between threads safely
  --> tests/ui/return_impl_future_nonsend.rs:18:13
   |
18 |     s.spawn(task()).unwrap();
   |             ^^^^^^ future created by async block is not `Send`
   |
   = help: within `impl Sized`, the trait `Send` is not implemented for `*mut ()`
note: captured value is not `Send`
  --> tests/ui/return_impl_future_nonsend.rs:13:24
   |
13 |         println!("{}", non_send as usize);
   |                        ^^^^^^^^ has type `*mut ()` which is not `Send`
note: required by a bound in `SendSpawner::spawn`
  --> src/spawner.rs
   |
   |     pub fn spawn<S: Send>(&self, token: SpawnToken<S>) -> Result<(), SpawnError> {
   |                     ^^^^ required by this bound in `SendSpawner::spawn`
